### PR TITLE
ARROW-8368: [C++][C Data Interface] Move several child arrays

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@toddlipcon @wesm This adds a test confirming that moving several child arrays works with the Arrow C++ exporter.

**Comment by wesm** [comment]
This seems acceptable to me, since it accounts for the use case where a consumer wishes to project a subset of fields from a producer where they may not be able to "push down" their subset request into the producer request. Should the format docs be amended to state that moving multiple children is permissible so long as the moves and then parent-release are handled as an "atomic surgery"?

**Comment by pitrou** [comment]
Yes, I should update the spec document.

File: docs/source/format/CDataInterface.rst [hunk 1]
```diff
@@ -620,10 +620,16 @@ the producer.
 As usual, the release callback will be called on the destination structure
 when it is not needed anymore.
 
-It is possible to move a child array, but the parent array MUST be released
-immediately afterwards, as it won't point to a valid child array anymore.
-This satisfies the use case of keeping only a subset of child arrays, while
-releasing the others.
+Moving child arrays
```
Comment (pitrou) [inline docs/source/format/CDataInterface.rst:623]: Does the wording below look good?

**Review by kszucs — APPROVED** [review]
LGTM

**Pull request closed without merging**